### DISCUSSION TITLE
MOOSE, Tutorial, error when using "make" in tutorial folder

### DISCUSSION THREAD

**Discussion by ananthanarasimhanj** [idaholab/moose#22066] [Q&A General] [answered]
I get the following error in double quotes when I try to use **make**. 

"
Linking Executable /home/ananth/projects/tutorial/tutorial-opt...
/home/ananth/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/ananth/projects/tutorial/lib/libtutorial-opt.so: undefined reference to `InputParameters validParams<Material>()'
collect2: error: ld returned 1 exit status
make: *** [/home/ananth/projects/moose/framework/app.mk:416: /home/ananth/projects/tutorial/tutorial-opt] Error 1
"

**Comment by GiudGiud** [comment]
Hello

You have likely forgotten to implement the validParams routine for an object. 

Either that or your objects were written at least 2 years ago and are using the old validParams templates syntax

Guillaume

**Reply by ananthanarasimhanj** [reply]
Thanks Guillaume. Where can i implement the validParams routing for an object? I am new to this, and following a video tutorial and I don't have this info from there.

**Reply by GiudGiud** [reply]
it should be in the source file. what object are you working on? are you starting from scratch or from someone's app?

**Reply by ananthanarasimhanj** [reply]
I will check. not the scratch. someone's.

**Reply by GiudGiud** [reply]
which app is that?
I d like to have a look to make sure I diagnosed the right issue

**Reply by ananthanarasimhanj** [reply]
I am using this github.com/shannon-lab/tutorial

Below is the video I am following, and I have timed the video link where i find error when i use make. 
https://youtu.be/GQUp718UxcY?t=900

I am also using moose in wsl.

**Reply by GiudGiud** [reply]
This repository is out of date
you need to do what is outlined in this post
https://mooseframework.inl.gov/newsletter/2020_04.html#!
to make it work again. 

@cticenhour is there an updated version of this tutorial

**Reply by cticenhour** [reply]
This tutorial was made for a specific plasma conference in 2018, so it would be very out-of-date at this point. @ananthanarasimhanj if you want to learn MOOSE in general, I would highly recommend that you check out the MOOSE examples and tutorials located at https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html. These can be done with a fresh copy of MOOSE cloned to your drive.

After that, if you are interested in MOOSE plus Zapdos plasma physics in light of the tutorial you found, I would direct you to https://github.com/shannon-lab/zapdos.

**Reply by ananthanarasimhanj** [reply]
@GiudGiud  @cticenhour  Thanks a lot for the inputs. I will follow the suggestions, and update here soon.

**Reply by ananthanarasimhanj** [reply]
@cticenhour  The examples in the link https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html is working for me.. Thank you for the inputs.. now I am trying to use the link https://shannon-lab.github.io/zapdos/getting_started/using_zapdos.html

In this link, i see the following instruction: After Zapdos is installed and tested, you should now be able to run input files using the zapdos-opt executable located at ~/projects/zapdos. Input files demonstrating the capabilities of Zapdos can be found in ~/projects/zapdos/tutorial and ~/projects/zapdos/tests. Any input file (say, one called example_input.i) can be run with the following basic syntax. ~/projects/zapdos/zapdos-opt -i example_input.i


Now, my question is, shouldnt I use "make" first, to  create -opt file? i dont see zapdos-opt within the tutorial or tests folder, where input files are there.

**Reply by cticenhour** [reply]
The Zapdos installation guide instructs you to build. So I would start here before progressing to the zapdos examples and tutorial: https://shannon-lab.github.io/zapdos/getting_started/installation.html

In short: yes, you should make Zapdos after pulling down the moose submodule for it.

**Reply by ananthanarasimhanj** [reply]
@cticenhour Thanks. I followed the installation page, and verified that its installed succefully. 

Following the successfully installation and verification, I given "make" in the zapdos directory. After that, if i wanted to run "diffusion-only.i" from "tutorial01-Diffusion" folder. I used the code "**zapdos-opt -i diffusion-only.i**" inside the directory "**turotrial01-Diffusion**". For this, i got the following message: "**zapdos-opt: command not found**"

**Reply by cticenhour** [reply]
`zapdos-opt` is not available in your system PATH, so that is why it cannot find the command. Since you are running an input within zapdos itself, it is better to refer to the executable directly, using relative file paths. If you are in the `tutorial01-Diffusion` folder, you would want to run:

```
../../zapdos-opt -i diffusion-only.i
```

**Reply by ananthanarasimhanj** [reply]
@cticenhour  Thanks it worked! I will proceed. Can I use paraview to view the out of tutorial01-Diffusion?

**Reply by cticenhour** [reply]
Yes, you should use Paraview to view the outputs of all of the zapdos tutorial problems.

**Reply by ananthanarasimhanj** [reply]
@cticenhour  when i try to use the command "git rebase origin/master" after updating zapdos, i get the following mesage.
error: cannot rebase: You have unstaged changes.
error: Please commit or stash them.

**Reply by cticenhour** [reply]
As the message says, you need to either stash your changes in a temporary location:
```
git stash
git rebase origin/master
git stash pop
```
where "pop" is the sub-command to bring back your changes, or you need to commit them:
```
git add <all_of_your_edited_files>
git commit -m "A message describing your changes"
git rebase origin/master
```
where you replace the fields I noted as necessary to get your files committed with a reasonable message of what those changes are. 

If you have further general questions about Zapdos usage like this, it is better suited for the Zapdos discussions forum: https://github.com/shannon-lab/zapdos/discussions

**Comment by cticenhour** [comment, marked as answer]
I would like to formally close this thread as "answered" since the original issue (out-of-date plasma tutorial code that should no longer be used) was addressed. Below is the relevant answer from the thread, with links to appropriate MOOSE and Zapdos references.

> This tutorial was made for a specific plasma conference in 2018, so it would be very out-of-date at this point. If you want to learn MOOSE in general, I would highly recommend that you check out the MOOSE examples and tutorials located at https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html. These can be done with a fresh copy of MOOSE cloned to your drive.
> 
> After that, if you are interested in MOOSE plus Zapdos plasma physics in light of the tutorial you found, I would direct you to https://github.com/shannon-lab/zapdos.

**Reply by ananthanarasimhanj** [reply]
Yes, please. @cticenhour .. Thanks a lot for your support.
